cassandra-executor/src/main/java/io/mesosphere/mesos/frameworks/cassandra/CassandraExecutor.java

```diff
@@ -124,19 +124,51 @@ public void launchTask(final ExecutorDriver driver, final TaskInfo task) {
                     driver.sendStatusUpdate(taskStatus(task, TaskState.TASK_RUNNING, details));
                     break;
                 case CASSANDRA_SERVER_RUN:
-                    safeShutdown();
+                    safeShutdown(driver);
                     serverTask = task;
                     serverTaskStatusOutstanding = true;
                     process = objectFactory.launchCassandraNodeTask(taskIdMarker, taskDetails.getCassandraServerRunTask());
                     jmxConnect = objectFactory.newJmxConnect(taskDetails.getCassandraServerRunTask().getJmx());
                     break;
                 case CASSANDRA_SERVER_SHUTDOWN:
-                    safeShutdown();
-                    driver.sendStatusUpdate(taskStatus(task.getExecutor().getExecutorId(),
-                            serverTask.getTaskId(),
+                    if (serverTask == null) {
+                        driver.sendStatusUpdate(slaveErrorDetails(task, "no server task", "-", SlaveErrorDetails.ErrorType.PROCESS_NOT_RUNNING));
+                    } else {
+                        TaskID serverTaskId = serverTask.getTaskId();
+                        if (jmxConnect != null) {
+                            LOGGER.info(taskIdMarker, "Stopping Cassandra Daemon...");
+                            try {
+                                try {
+                                    jmxConnect.getStorageServiceProxy().stopDaemon();
+                                } catch (Throwable ignored) {
+                                    // any kind of strange exception may occur since shutdown closes everything
+                                }
+                                // TODO make shutdown timeout configurable
+                                long timeoutAt = System.currentTimeMillis() + 60000L;
+                                while (true) {
+                                    if (timeoutAt < System.currentTimeMillis()) {
+                                        // TODO push some more information to scheduler if regular shutdown failed
+                                        break;
+                                    }
+                                    try {
+                                        int exitCode = process.exitValue();
+                                        LOGGER.info(taskIdMarker, "Cassandra process terminated with exit code {}", exitCode);
+                                        break;
+                                    } catch (IllegalThreadStateException e) {
+                                        // ignore
+                                    }
+                                }
+                            } catch (Exception e) {
+                                LOGGER.error(taskIdMarker, "Failed to stop Cassandra daemon - forcing process destroy", e);
+                            }
+                        }
```

Comment:
we probably want to add an `else` here that will fall back to trying to run `kill -9 $(cat cassandra.pid)` if we're not able to use the jmx `stopDaemon` method.

We should also remove `cassandra.pid` when we shutdown the server.